### ISSUE TITLE
1.1.0-stable build break on MacOS: non-portable `BIO_PRI64`

### ISSUE THREAD

**Issue by vdukhovni** [openssl/openssl#4747]
To get 1.1.0 to build on MacOS (10.13 High Sierra) I had to change `BIO_PRI64` from `"l"` to `"ll"`:
```
diff --git a/e_os.h b/e_os.h
index dfa159ff9a..66314c9f71 100644
--- a/e_os.h
+++ b/e_os.h
@@ -32,9 +32,9 @@ extern "C" {
 /*
  * BIO_printf format modifier for [u]int64_t.
  */
 # if defined(__LP64__) || (defined(__SIZEOF_LONG__) && __SIZEOF_LONG__==8)
-#  define BIO_PRI64 "l"     /* 'll' does work "universally", but 'l' is
+#  define BIO_PRI64 "ll"     /* 'll' does work "universally", but 'l' is
                              * here to shut -Wformat warnings in LP64... */
 # else
 #  define BIO_PRI64 "ll"
 # endif
```

Otherwise I get:
```
cc  -I. -Icrypto/include -Iinclude -I../openssl -I../openssl/crypto/include -I../openssl/include -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/opt/openssl/1.1.0/ssl\"" -DENGINESDIR="\"/opt/openssl/1.1.0/lib/engines-1.1\"" -O3 -D_REENTRANT -arch x86_64 -DL_ENDIAN -Wall  -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wsign-compare -Wmissing-prototypes -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Qunused-arguments -Wno-language-extension-token -Wno-extended-offsetof -Wconditional-uninitialized -Wincompatible-pointer-types-discards-qualifiers -Wmissing-variable-declarations -fPIC -MMD -MF crypto/asn1/x_int64.d.tmp -MT crypto/asn1/x_int64.o -c -o crypto/asn1/x_int64.o ../openssl/crypto/asn1/x_int64.c
../openssl/crypto/asn1/x_int64.c:105:51: error: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long')
      [-Werror,-Wformat]
        return BIO_printf(out, "%"BIO_PRI64"d\n", **(int64_t **)pval);
                                ~~~~~~~~~~~~~     ^~~~~~~~~~~~~~~~~~
                                %lld
../openssl/crypto/asn1/x_int64.c:106:47: error: format specifies type 'unsigned long' but the argument has type 'uint64_t'
      (aka 'unsigned long long') [-Werror,-Wformat]
    return BIO_printf(out, "%"BIO_PRI64"u\n", **(uint64_t **)pval);
                            ~~~~~~~~~~~~~     ^~~~~~~~~~~~~~~~~~~
                            %llu
2 errors generated.
make[1]: *** [crypto/asn1/x_int64.o] Error 1
make: *** [all] Error 2
```
The Makefile generated by Configure shows:
```
PLATFORM=darwin64-x86_64-cc
OPTIONS=--strict-warnings --prefix=/opt/openssl/1.1.0 enable-shared no-asan no-crypto-mdebug no-crypto-mdebug-backtrace no-ec_nistp_64_gcc_128 no-egd no-fuzz-afl no-fuzz-libfuzzer no-heartbeats no-md2 no-msan no-rc5 no-sctp no-ssl-trace no-ssl3 no-ssl3-method no-ubsan no-unit-test no-weak-ssl-ciphers no-zlib no-zlib-dynamic
CONFIGURE_ARGS=("--strict-warnings", "--prefix=/opt/openssl/1.1.0", "shared", "darwin64-x86_64-cc")
SRCDIR=../openssl
BLDDIR=.
```

**Comment by levitte** [comment]
That's quite odd...  it means that the compiler defines `__LP64__` or `__SIZEOF_LONG__=8`, but still has an unsigned long that's 32 bit?

**Comment by vdukhovni** [comment]
I think it's not so much that unsigned long is 32-bit (it is not), but rather than it expects "%lld" for int64_t even when long is also 64-bit.  Which is an annoying warning, but there it is.  The point is that "%lld" works for both 32-bit code and 64-bit code, so that's what the compiler wants to see. :-(

**Comment by vdukhovni** [comment]
One way out of this is to cast the `uint64_t` to a `uintmax_t`, and use "%jd", on any platform that has those.  Or even better use `PRIu64` from `<inttypes.h>` when that's available. http://pubs.opengroup.org/onlinepubs/009695399/basedefs/inttypes.h.html

**Comment by levitte** [comment]
> I think it's not so much that unsigned long is 32-bit (it is not), but rather that it expects "%lld" for int64_t even when long is also 64-bit. Which is an annoying warning, but there it is. The point is that "%lld" works for both 32-bit code and 64-bit code, so that's what the compiler wants to see. :-(

Well, it's not quite that simple, and reading the comment that's right there on the line you're changing should give you a bit of a hint.  So here's what happens with gcc (on Linux if that matters) after your change:

```
gcc  -I. -Icrypto/include -Iinclude -I../1.1.0 -I../1.1.0/crypto/include -I../1.1.0/include -DDSO_DLFCN -DHAVE_DLFCN_H -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-1.1\"" -Wall -O0 -g -pthread -m64 -DL_ENDIAN  -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wsign-compare -Wmissing-prototypes -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wa,--noexecstack -fPIC -DOPENSSL_USE_NODELETE -MMD -MF crypto/asn1/x_int64.d.tmp -MT crypto/asn1/x_int64.o -c -o crypto/asn1/x_int64.o ../1.1.0/crypto/asn1/x_int64.c
../1.1.0/crypto/asn1/x_int64.c: In function 'uint64_print':
../1.1.0/crypto/asn1/x_int64.c:105:32: error: format '%lld' expects argument of type 'long long int', but argument 3 has type 'int64_t {aka long int}' [-Werror=format=]
         return BIO_printf(out, "%"BIO_PRI64"d\n", **(int64_t **)pval);
                                ^~~                ~~~~~~~~~~~~~~~~~~
../1.1.0/crypto/asn1/x_int64.c:105:45: note: format string is defined here
         return BIO_printf(out, "%"BIO_PRI64"d\n", **(int64_t **)pval);
                                 ~~~~~~~~~~~~^
../1.1.0/crypto/asn1/x_int64.c:106:28: error: format '%llu' expects argument of type 'long long unsigned int', but argument 3 has type 'uint64_t {aka long unsigned int}' [-Werror=format=]
     return BIO_printf(out, "%"BIO_PRI64"u\n", **(uint64_t **)pval);
                            ^~~                ~~~~~~~~~~~~~~~~~~~
../1.1.0/crypto/asn1/x_int64.c:106:41: note: format string is defined here
     return BIO_printf(out, "%"BIO_PRI64"u\n", **(uint64_t **)pval);
                             ~~~~~~~~~~~~^
cc1: all warnings being treated as errors
Makefile:1232: recipe for target 'crypto/asn1/x_int64.o' failed
make[1]: *** [crypto/asn1/x_int64.o] Error 1
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
Makefile:130: recipe for target 'all' failed
make: *** [all] Error 2
```

(before you ask, clang has a similar output)

> One way out of this is to cast the uint64_t to a uintmax_t, and use "%jd", on any platform that has those.

You forget that this is about `BIO_printf`, i.e. *our* implementation.  "%jd" isn't present in 1.1.0.  It *is*, however, present in 1.1.1-dev.  It was deemed a new feature that appeared after 1.1.0 was released, so it didn't sneak back into 1.1.0 ('cause policy).

> Or even better use PRIu64 from <inttypes.h> when that's available.

I'll let the commit that replaced PRIu64 with BIO_PRI64 speak for itself: 70b9063cd24904fd3d0b83173c51fdcf34085967

So to sum it up, your change is problematic...  *However*, if you look into the macros made by your compiler, you might find what you need to avoid the "l" definition.

**Comment by vdukhovni** [comment]
I tried to make it clear that my change was just a quick hack to get a built, it was not a suggested fix. I guess that was easy to miss.  Sorry about that.  As for the macros defined by the compiler, potentially relevant are:

```
#define _LP64 1
#define __APPLE_CC__ 6000
#define __APPLE__ 1
#define __INT64_C_SUFFIX__ LL
#define __INT64_FMTd__ "lld"
#define __INT64_FMTi__ "lli"
#define __INT64_MAX__ 9223372036854775807LL
#define __INT64_TYPE__ long long int
#define __LP64__ 1
#define __UINT64_FMTX__ "llX"
#define __UINT64_FMTo__ "llo"
#define __UINT64_FMTu__ "llu"
#define __UINT64_FMTx__ "llx"
#define __UINT64_MAX__ 18446744073709551615ULL
#define __UINT64_TYPE__ long long unsigned int
#define __amd64 1
#define __amd64__ 1
#define __apple_build_version__ 9000038
#define __clang__ 1
#define __clang_major__ 9
#define __clang_minor__ 0
#define __clang_patchlevel__ 0
#define __x86_64 1
#define __x86_64__ 1
```

Perhaps some of these are sufficient to pick out a combination that will not break Linux.

**Comment by dot-asm** [comment]
This is the kind of game one can't win. Note that commentary even says that "ll" does work universally, and that "l" is there to shut up compiler warnings. As for using some macro, well, it's not given that it works. Compiler is obliged to provide macro that matches its run-time, and what we are looking at here is BIO_printf, i.e. function that *we* implement. As an example consider Microsoft where macro in question would have to be "I64", but BIO_printf doesn't understand it. Yes, one can solve it with excessive #ifdef spaghetti, but question is how far is one ready to go [to meed contradicting requirements]? When you can as well configure with --strict-warnings -Wno-format (or without --strict-warnings). Also note that in master there is no BIO_PRI64 anymore. For the above mentioned reasons...

**Comment by dot-asm** [comment]
> configure with --strict-warnings -Wno-format (or without --strict-warnings)

This might appear controversial, but as just said, we do face *contradicting* requirements, which also change all the time, so that advice to customize strict warnings (or even occasionally avoid) is not actually unreasonable. Because it is a trade-off. Indeed, --strict-warnings are there to warn us about bad coding, but we do choose to *not* pay attention to some of them by consciously suppressing, e.g. -Wno-missing-field-initializers. So how is additional -Wno-format different?

**Comment by mattcaswell** [comment]
What should happen with this issue now? It's currently against the 1.1.1 milestone...

**Comment by levitte** [comment]
@vdukhovni, is this still an issue?

**Comment by vdukhovni** [comment]
I failed to reproduce this with a recent build, so closing for now.
